enhancements/olm-cmd-workflows.md

```diff
@@ -0,0 +1,165 @@
+---
+title: README describing new command workflows for OLM enabled operators
+authors:
+  - "@varshaprasad96"
+reviewers:
+  - "@jlanford"
+  - "@dmesser"
+  - "@estroz"
+approvers:
+  - "@jlanford"
+  - "@dmesser"
+  - "@estroz"
+creation-date: 2020-04-09
+last-updated: 2020-04-09
+status: provisional
+---
+
+# README describing new command workflows for OLM enabled operators
+
+## Release Signoff Checklist
+
+- \[ \] Enhancement is `provisional`
+- \[ \] Design details are appropriately documented from clear requirements
+- \[ \] Test plan is defined
+- \[ \] Graduation criteria for dev preview, tech preview, GA
+- \[ \] User-facing documentation is created.
+
+## Summary
+
+This proposal provides documentation describing the possible workflows of new commands for generating operator manifests with user inputs and operator bundle for OLM to deploy the operator.
+
+We plan to improve the developer experience for operator authors using Operator SDK with OLM by:
+1. Providing an interactive subcommand to get inputs for generating CSV.
+2. Providing a single command for generating operator manifests and operator bundle image.
+
+## Motivation
+
+Currently, the `operator-sdk generate csv` command is used for generating CSVs, which scaffolds the project and writes the CSV manifest on disk. This process still requires the developers to manually edit the CSV when any of the required fields are missing. Also, with the deprecation in the use of package manifests for OLM, generating CSVs and creating operator bundle does not require two separate commands. Integrating both the functionalities and providing a single command with an interactive input for collecting the required UI metadata to generate CSV and operator bundle will result in a better developer experience.
+
+## Goals
+
+The proposal is aimed at explaining the resulting process of developing an OLM enabled operator from scratch with the above mentioned improvements.
+
+## Proposal
+
+### README
+
+The following document discusses workflows for creating a new operator and running it with OLM:
+
+Operator creation workflow:
+
+1. Create a new operator project using the SDK Command Line Interface(CLI)
+2. Define new resource APIs by adding Custom Resource Definitions(CRD)
+3. Define Controllers to watch and reconcile resources
+4. Write the reconciling logic for your Controller using the SDK and controller-runtime APIs
+5. Use the SDK CLI to build and generate the operator deployment manifests
+
+OLM Workflow:
+
+1. Use `operator-sdk generate bundle` to generate kustomize base templates with CSV metadata and Dockerfile to build the operator.
```

Comment:
@varshaprasad96 @dmesser a few thoughts on the base project -> OLM-ready workflow:

Eventually we'll want to move bundle (manifest) generation to `kustomize` to align with kubebuilder-style deployment. This will be especially relevant when CSV-less bundles come around. Therefore I don't thing we should build `generate bundle` so that it both ingests manifests from disk/stdin AND generates UI metadata. There should be a step, perhaps in `operator-sdk init`, that scaffolds UI metadata in a [base](https://github.com/kubernetes-sigs/kustomize/blob/master/docs/glossary.md#base). The user then updates it manually or via an interactive prompt, after which they can run `kustomize build config/olm-catalog` to generate their bundle to disk or as input to `bundle create` or `run --olm`.

The workflow then becomes:
```console
$ operator-sdk init
Interactive metadata prompt. Please fill in the following UI fields:
...
$ operator-sdk create api --group cache --version v1 --kind Memcached
...
```
Then:
```console
$ make manifests-olm
kustomize build config/olm-catalog | operator-sdk generate bundle --output-dir config/olm-catalog/manifests
```
or
```console
$ make bundle IMG=quay.io/example/memcached-operator:v0.0.1
kustomize build config/olm-catalog | operator-sdk create bundle quay.io/example/memcached-operator:v0.0.1
```
or
```console
$ make run-olm
kustomize build config/olm-catalog | operator-sdk run --olm
```

The make recipes above are just suggestions to illustrate how moving to `kustomize` will be much more flexible. I have something a little different set up in https://github.com/operator-framework/operator-sdk/pull/2860 because we still need to regenerate CSV definitions from API types. Still tinkering around with that PR. Thoughts?

Reply:
Regardless on-disk manifests in a kubebuilder project are mostly just bases, so _some_ part of bundle generation needs to be run through `kustomize`. We have two options then, which are inverses:

1. Run `kustomize | operator-sdk`, which is how kubebuilder does things.
1. Invoke `kustomize` within `operator-sdk`.

The former is a lot more appealing to me.